crossbar/worker/proxy.py

```diff
@@ -126,248 +203,280 @@ def onMessage(self, msg):
-        to our backend.
-        """
-        if self._session_id is not None:
-            raise ProtocolError("Hello received but session established already")
-
-        self._realm = msg.realm
-        self._session_id = util.id()
-        self._goodbye_sent = False
-
-        authid = msg.authid
-        authrole = msg.authrole or 'anonymous'
-        # XXX x_cb_node_id, x_cb_peer, x_cb_pid
+            else:
+                raise ProtocolError("Received {} message while proxy frontend session is not joined".format(msg.__class__.__name__))
 
-        details = types.HelloDetails(
-            realm=self._realm,
-            authmethods=msg.authmethods,
-            authid=authid,
-            authrole=authrole or 'anonymous',
-            authextra=msg.authextra,
-            session_roles=msg.roles,
-            pending_session=self._pending_session_id
-        )
+        else:
+            # frontend session is established: process WAMP message
 
-        res = self._do_authentication(self._realm, details)
-
-        # Note: "roles" come from self._router.attach() in non-proxy
-        # code
-        roles = {
-            'broker': RoleBrokerFeatures(
-                publisher_identification=True,
-                pattern_based_subscription=True,
-                session_meta_api=True,
-                subscription_meta_api=True,
-                subscriber_blackwhite_listing=True,
-                publisher_exclusion=True,
-                subscription_revocation=True,
-                event_retention=True,
-                payload_transparency=True,
-                payload_encryption_cryptobox=True,
-            ),
-            'dealer': RoleDealerFeatures(
-                caller_identification=True,
-                pattern_based_registration=True,
-                session_meta_api=True,
-                registration_meta_api=True,
-                shared_registration=True,
-                progressive_call_results=True,
-                registration_revocation=True,
-                payload_transparency=True,
-                testament_meta_api=True,
-                payload_encryption_cryptobox=True,
-                call_canceling=True,
-            )
-        }
-
-        msg = None
-        if isinstance(res, types.Accept):
-            msg = message.Welcome(
-                self._session_id,
-                roles,
-                realm=self._realm,
-                authid=self._authid,
-                authrole=self._authrole,
-                authmethod=self._authmethod,
-                authprovider=self._authprovider,
-                authextra=self._authextra,
-                #  custom=custom,
-            )
+            if isinstance(msg, message.Hello) or isinstance(msg, message.Abort) or isinstance(msg, message.Authenticate):
+                raise ProtocolError("Received {} message while proxy frontend session is already joined".format(msg.__class__.__name__))
 
```

Comment:
Is Abort only during the handshake phase?

Reply:
yes: once a session is established, GOODBYE is the only message allowed to initiate the WAMP closing handshake (well, and the handshake is merely a GOODBYE reply from the peer) - also https://wamp-proto.org/_static/gen/wamp_latest.html#difference-between-abort-and-goodbye